### PR TITLE
[WiP] Use Machinomy 1.7.9

### PR DESCRIPTION
**Pull request by michielbdejong**

### CODE REVIEW

**Comment by michielbdejong** [comment]
machinomy:AbstractChannelsDatabase Spending channel with ID 0xf4edbf1860964ef7bb2b84910db3807b00000000000000000000000000000000 +1s
(node:12604) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Invalid signature length
(node:12604) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by michielbdejong** [comment]
@emschwartz the server is now saying:
```sh
GOT PAYMENT { channelId: '0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000',
  value: '30000',
  sender: '0x3284166218919725e0d27da8dc54b9e8c2e3c31b',
  receiver: '0xA1029EBA3E863423E324E8AE8851E91598435505',
  price: '10000',
  channelValue: '32000',
  v: 28,
  r: '0x46298e4c344abc114cdb0461a8534a77bae8ad9bc67d9a0184a535e63779381c',
  s: '0x1a2102d1772733e48e8a519db30540301e915c14164a13d974a7fe487baabaab',
  meta: '' }
  machinomy:Client Received payment request. Sender: 0x3284166218919725e0d27da8dc54b9e8c2e3c31b / Receiver: 0xA1029EBA3E863423E324E8AE8851E91598435505 +0ms
  machinomy:ChannelManager Queueing payment of 10000 Wei to channel with ID 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000. +0ms
  machinomy:ChannelManager Adding 10000 Wei to channel with ID 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000. +3ms
  machinomy:ChannelContract Fetching settlement period for channel 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000 +0ms
  machinomy:ChannelManager Received invalid payment from 0x3284166218919725e0d27da8dc54b9e8c2e3c31b! +1s
  ilp-plugin-mini-accounts btp packet not accepted Error: Invalid payment.
    at mutex.synchronizeOn (/Users/michiel/gh/interledgerjs/ilp-plugin-ethereum-asym-server/node_modules/machinomy/dist/lib/channel_manager.js:83:19)
    at <anonymous> +1s

**Comment by michielbdejong** [comment]
combine this PR with https://github.com/interledgerjs/ilp-plugin-ethereum-asym-server/pull/7

**Comment by michielbdejong** [comment]
i did notice that https://github.com/machinomy/machinomy/blob/master/lib/payment_manager.ts#L41 is failing, with spent=0, price=100000, value=200000. also canClaim is failing, but that's calculated by the deployed contract, so hard to debug.

File: index.js [hunk 1]
```diff
@@ -114,7 +115,7 @@ class Plugin extends PluginBtp {
     const primary = data.protocolData[0]
     if (primary.protocolName === 'machinomy') {
       const payment = new Payment(JSON.parse(primary.data.toString()))
```
Comment (sentientwaffle) [inline index.js:117]: `Payment` isn't require()d anymore.

Comment (michielbdejong) [reply index.js:117]: Thanks! It's not reaching that code yet, but will look into that when the sending of the payment works.

**Comment by michielbdejong** [comment]
Apparently the Machinomy contract on Rinkeby should be the right one for 1.7.9, but we're still running into invalid signature errors. My current approach is going to be to rewrite https://github.com/machinomy/machinomy/blob/master/examples/machinomy.ts to use Infura, see what effect `machinomy.buy` has in that example, and try to work the rest out from there.

**Comment by michielbdejong** [comment]
I think we need to use https://machinomy.com/documentation/api/classes/_index_.machinomy.html#deposit instead of nextPayment / payment.

**Comment by michielbdejong** [comment]
a wait, deposit is used on https://github.com/interledgerjs/ilp-plugin-ethereum-asym-client/blob/master/index.js#L90, so that's not it. But still, there is neither a `payment` nor a `nextPayment` in the currents docs. So then probably we need to replace `machinomy.payment` with [`machinomy.buy`](https://machinomy.com/documentation/api/classes/_index_.machinomy.html#buy)

**Comment by michielbdejong** [comment]
aaah ok, i think i'm getting closer to the truth now. So the way we're using Machinomy is not actually officially supported. You're supposed to 'buy' a URL, and then it opaquely takes care of sending the claim update over http. So we want to construct the payment, and then not send it over http. So to debug that, I'll set up a mock http server, and use that as the gateway in a buy operation, and then see what it posts, and how it gets to create the payment object for that, and then i should be able to compare that with what we're doing in our code.

**Comment by michielbdejong** [comment]
ok, that seems to have successfully created a payment object, now i just have to trace back how it did it internally:
```sh
ilp-plugin-ethereum-asym-client funding channel for 512000 +945ms
  machinomy:ChannelContract Fetching state for channel 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000 +945ms
  machinomy:ChannelContract Depositing 512000 into channel 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000 +1s
  machinomy:ChannelContract Fetching state for channel 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000 +32s
  machinomy:ChannelContract Fetching state for channel 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000 +1s
  machinomy:ChannelContract Fetching state for channel 0xf4edbf1860964ef7bb2b84910db3807b00000000000000000000000000000000 +0ms
  machinomy:ChannelContract Fetching state for channel 0xdb24e0944b3c47a08ea1375a5e66c85000000000000000000000000000000000 +981ms
  machinomy:Client Attempting to send payment to http://localhost:3000. Sender: 0x3284166218919725e0d27da8dc54b9e8c2e3c31b / Receiver: 0xA1029EBA3E863423E324E8AE8851E91598435505 / Amount: 100000 +0ms

**Comment by emschwartz** [comment]
Sergey put in the payment thing for this use case because we agreed it made more sense to send it over BTP instead of HTTP. Trying it out with HTTP to see if it has the same error sounds good though

**Comment by michielbdejong** [comment]
I compiled https://github.com/machinomy/machinomy/blob/master/examples/client.ts and https://github.com/machinomy/machinomy/blob/master/examples/server.ts and made them use my Rinkeby test accounts via infura instead of Ropsten test accounts via geth. And then I'm getting the same error. So looks like it's not related to our code. Next step: try those examples the way they're meant to be used, with a local geth instance synced to Ropsten.

**Comment by michielbdejong** [comment]
Actually, Ropsten via Infura works! No longer getting the InvalidPayment error now.

Next error we're running into is this one:
```sh
  machinomy:Client Received payment request. Sender: 0x3284166218919725e0d27da8dc54b9e8c2e3c31b / Receiver: 0xa1029eba3e863423e324e8ae8851e91598435505 +0ms
  machinomy:ChannelManager Queueing payment of 1000 Wei to channel with ID 0x5283d1f329f64877af655a5280119c3500000000000000000000000000000000. +0ms
  machinomy:ChannelManager Adding 1000 Wei to channel with ID 0x5283d1f329f64877af655a5280119c3500000000000000000000000000000000. +3ms
equals? 0 1000 1000
  machinomy:ChannelContract Fetching settlement period for channel 0x5283d1f329f64877af655a5280119c3500000000000000000000000000000000 +0ms
(node:40687) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: Cannot fetch settlement period for non-existent channel 0x5283d1f329f64877af655a5280119c3500000000000000000000000000000000.
(node:40687) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by michielbdejong** [comment]
Getting that error consistently when using Infura. Trying to sync `geth --testnet` now so that I can really run Machinomy's example scripts as they are meant to be run, but it's slow.

**Comment by michielbdejong** [comment]
Now running into this error:

```sh
Error: Unidirectional has not been deployed to detected network (network/artifact mismatch)
    at /Users/michiel/gh/machinomy/machinomy/node_modules/truffle-contract/contract.js:455:17
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
```

with:

```js
self._json.networks == { '4':
   { events: {},
     links: { ECRecovery: '0x4ed0ef37c7746a094e7cbb1eb40791e074ac258b' },
     address: '0x6b77c2b4f5ba976baed454ff2a8aba5e60c6705f',
     transactionHash: '0xcfcf22404b81d3f5dec4178b670a0b7cac2098f825477c5268c0422776910270' } }
```
and
```js
self.network_id  == '3'
```

**Comment by michielbdejong** [comment]
According to https://ethereum.stackexchange.com/questions/17051/how-to-select-a-network-id-or-is-there-a-list-of-network-ids, 3 is Ropsten and 4 is Rinkeby.

`self._json.networks` is a wrong version (it had installed machinomy-contracts v4.0.6 on my machine), compared to https://github.com/machinomy/machinomy-contracts/blob/master/build/contracts/Unidirectional.json#L13453 (v4.0.9 is the latest)

**Comment by michielbdejong** [comment]
Fixed that in https://github.com/machinomy/machinomy/pull/193 and now I'm back at the 'Invalid payment' problem, but on Ropsten this time...

Same if I use machinomy-contracts v4.0.5 instead of v4.0.9...

@ukstv I think I reached the point where I really need your help :) Maybe if you could try to reproduce https://github.com/machinomy/machinomy/issues/194 then we can see if the issue is related to the use of Infura?

**Pull request merged**